client/client.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2019 Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	jaegerpb "github.com/jaegertracing/jaeger/model"
+	"go.opencensus.io/stats/view"
+)
+
+const (
+	idleConnTimeout     = 30 * time.Second
+	tlsHandshakeTimeout = 10 * time.Second
+	dialerTimeout       = 3 * time.Second
+	dialerKeepAlive     = 30 * time.Second
+
+	// default values
+	defaultNumWorkers  uint64 = 8
+	defaultMaxRetries  uint64 = 8
+	defaultMaxIdleCons        = 100
+	defaultHTTPTimeout        = 5 * time.Second
+)
+
+type sendRequest struct {
+	message []byte
+	spans   int64
+	batches int64
+}
+
+// Client implements an HTTP sender for the SAPM protocol
+type Client struct {
+	numWorkers  uint64
+	maxIdleCons uint64
+	maxRetries  uint64
+	endpoint    string
+	authToken   string
+	httpClient  *http.Client
+
+	workers chan *worker
+}
+
+// New creates a new SAPM Client
+func New(opts ...Option) (*Client, error) {
+	views := metricViews()
+	if err := view.Register(views...); err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		numWorkers:  defaultNumWorkers,
+		maxRetries:  defaultMaxRetries,
+		maxIdleCons: defaultMaxIdleCons,
+	}
+
+	for _, opt := range opts {
+		err := opt(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if c.endpoint == "" {
+		return nil, fmt.Errorf(
+			"endpoint cannot be empty. WithEndpoint option must be called with a valid endpoint value",
+		)
+	}
+
+	if c.httpClient == nil {
+		c.httpClient = &http.Client{
+			Timeout: defaultHTTPTimeout,
+			Transport: &http.Transport{
+				Proxy: http.ProxyFromEnvironment,
+				DialContext: (&net.Dialer{
+					Timeout:   dialerTimeout,
+					KeepAlive: dialerKeepAlive,
+				}).DialContext,
+				MaxIdleConns:        int(c.maxIdleCons),
+				MaxIdleConnsPerHost: int(c.maxIdleCons),
+				IdleConnTimeout:     idleConnTimeout,
+				TLSHandshakeTimeout: tlsHandshakeTimeout,
+			},
+		}
+	}
+
+	c.workers = make(chan *worker, c.numWorkers)
+	for i := uint64(0); i < c.numWorkers; i++ {
+		w := newWorker(c.httpClient, c.endpoint, c.authToken, c.maxRetries)
+		c.workers <- w
+	}
+
+	return c, nil
+}
+
+// Export takes a Jaeger batch and uses one of the available workers to export it synchronously.
+// It internally retries with a back-off in case of the server responding with HTTP 429.
+func (sa *Client) Export(ctx context.Context, batch *jaegerpb.Batch) error {
```

Comment:
I'm not sure we need a worker pool in the client?  It looks like the work and (potential retries) are all done on the calling routine anyway.  We could move the worker methods into this client.  The worker pool could be very useful if we want to have a dedicated routine doing the work asynchronously from the calling routine.

Reply:
Yes, this will block if all workers are busy.

@codesmith14 I think this is needed for high throughput. All other Otel exporters I've written and seen have done this to increase their throughput.

Reply:
I don't think i understand how this increases throughput? The work is done on the calling routine.  Nothing is done asynchronously outside of the calling routine.  I think the only thing the pool is doing is throttling the number of concurrent requests using the http.Client?

Reply:
At best it seems as though this is a dedicated pool of pre-allocated workers that reduces GC pressure very slightly by avoiding reallocation of workers in the `newWorker` function.  It doesn't really seem to do anything to increase throughput since `w.export` is blocking for the full RTT of the request (the GC optimization is probably negligible).

Reply:
The client itself can be called concurrently (by collector) and it'll be able to export as many request concurrently as there are workers. Without workers the client will only export one request at a time. Another alternative is to get rid of workers and do one goroutine per request but in my experience that can cause further slowdown under high load because there is no cap on number of goroutines and it causes goroutine contention. This is just a way to limit the number of goroutines we use to export concurrently.  There is a test that sort of  proves speed up when caller has multipe requests to export at the same time. https://github.com/signalfx/sapm-proto/pull/8/files#diff-fce84029d346504e44323c30d21a3808R177